### ISSUE TITLE
autodoc_typehints='signature' doesn't handle None

### ISSUE THREAD

**Issue by godlygeek** [sphinx-doc/sphinx#7428]
**Describe the bug**
Given a Python file containing:

```py
def foo() -> None: pass
```

And a `conf.py` containing:
```py
extensions = ["sphinx.ext.autodoc", "sphinx.ext.intersphinx"]
intersphinx_mapping = {"python": ("https://docs.python.org/3", None)}
```

The latest Sphinx installed from PyPI generates:

```rst
.. py:function:: foo() -> None
   :module: type_hint_test
```

And then winds up generating a warning, if `-n` is enabled:

```
type_hint_test.py:docstring of type_hint_test.foo::
WARNING: py:class reference target not found: None
```

Which of course fails the build if `-W` is enabled.

This works fine if the return type is `int` instead of `None` - I'm guessing that the issue is the type is resolved using `py:class`, but that `None` is special, as per [PEP-484](https://www.python.org/dev/peps/pep-0484/#using-none):

> When used in a type hint, the expression `None` is considered equivalent to `type(None)`.

**To Reproduce**
```sh
mkdir -p sphinx_type_hint_none
cd sphinx_type_hint_none

printf "%s\n" > type_hint_test.py \
    "def foo() -> None: pass"     \

mkdir -p docs

printf "%s\n" > docs/conf.py                                                \
    'extensions = ["sphinx.ext.autodoc", "sphinx.ext.intersphinx"]'         \
    'intersphinx_mapping = {"python": ("https://docs.python.org/3", None)}' \

printf "%s\n" > docs/index.rst       \
    '.. automodule:: type_hint_test' \
    '.. autofunction:: foo'          \

mkdir -p html
python3.8 -m sphinx -vvv -nW -b html --keep-going docs html
```

**Expected behavior**
Documentation generation completes without an error, and includes a link to https://docs.python.org/3/library/constants.html#None

**Environment info**
- OS: Linux 4.4.0
- Python version: 3.8.1
- Sphinx version: 3.0.0
- Sphinx extensions:  sphinx.ext.autodoc, sphinx.ext.intersphinx

**Additional context**
Appears to have been introduced in 3.0.0; our CI had no issue with the latest version on PyPI as of Friday.

**Comment by tk0miya** [comment]
+1; Reasonable.
I worked to suppress warnings for built-in types in 3.0 (see #6895). But I forgot to handle `None`.

**Comment by godlygeek** [comment]
Note that if I use `autodoc_typehints='description'`, the generated HTML contains:

```html
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)">None</a></p>
</dd>
```

That is, it generates a clickable link to the `None` singleton from the `Return type` section.  It seems inconsistent for `autodoc_typehints='description'` to generate a link to the None singleton's documentation but for `autodoc_typehints='signature'` to not generate a link.

**Comment by tk0miya** [comment]
Fixed by #7432 
Thank you for reporting.

@godlygeek I think it is another topic. So please file another issue. Thanks,
